### ISSUE TITLE
[FEATURE-REQUEST] Bigger than memory Random Forest - with starter code

### ISSUE THREAD

**Issue by Chris-hughes10** [vaexio/vaex#1262]
A Random Forest model can be a valuable tool during initial data exploration and feature selection - an area that vaex excels at - but most available implementations are constrained such that the data must fit into memory; whilst sampling can address this, as datasets continue to grow, this is becoming increasingly common.. Whilst vaex ml provides a wrapper which enables the use of the scikit learn implementation, as this is a simple wrapper, this first loads all of the data into memory.

For a personal project, I experimented with this, and have found that with relatively little effort, the scikit implementation can be adapted to use vaex, and can be used on larger than memory datasets - I would be happy to share the code for this if there is interest.

**Comment by JovanVeljanoski** [comment]
Hi @Chris-hughes10 

Yeah please do that would be awesome! You can also submit a PR if you'd like!

We are also looking to collaborate with projects such as River (see #1256, https://github.com/online-ml/river/discussions/515), which implement online-versions of many algorithms. They have things like decision trees and ensembles, and other interesting models. 

While most of their models train/predict only one sample at a time (which is too slow for big data), some do implement mini-batch learning / predicting which can be super fast! I hope they will implement batch learning for more of their models at some point soon.

**Comment by Chris-hughes10** [comment]
I would be open to collaborating on a PR, I had a quick look through the codebase and it seems like there is a lot going on for someone largely unfamiliar with it! I don't have a huge amount of time though unfortunately!

This is the code I have so far, it is still pretty rough, but it works. Essentially I went through scikit's implementation and tried to change as little as I could. Where I had to make changes, at present, I have replaced the function with my own version which is post-fixed with `_patch`. This is for clarity over what the original function was, and I wouldn't expect it to remain this way;; as I used the original functions as a base, there is a lot of similarity in some areas! I didn't implement OOB, sample weights or balanced subsampling, as I didn't really have a need for these when using a larger dataset.

The modified functions are:
```
from warnings import warn

import numpy as np
from joblib import Parallel, delayed
from sklearn.ensemble._forest import (
    _generate_sample_indices,
    _get_n_samples_bootstrap,
    MAX_INT,
)
from sklearn.metrics import accuracy_score
from sklearn.utils.fixes import _joblib_parallel_args
from sklearn.utils.validation import _check_sample_weight, check_random_state


def fit_patched(self, X, y, sample_weight=None):

    if sample_weight is not None:
        sample_weight = _check_sample_weight(sample_weight, X)

    # Remap output
    self.n_features_ = X.shape[1]
    self.n_outputs_ = y.shape[1]

    y, class_weights_df = _validate_y_class_weight_patched(self, y)

    if class_weights_df is not None:
        sample_weight = class_weights_df

    # Get bootstrap sample size
    n_samples_bootstrap = _get_n_samples_bootstrap(
        n_samples=X.shape[0], max_samples=self.max_samples
    )

    # Check parameters
    self._validate_estimator()

    if not self.bootstrap and self.oob_score:
        raise ValueError("Out of bag estimation only available" " if bootstrap=True")

    random_state = check_random_state(self.random_state)

    if not self.warm_start or not hasattr(self, "estimators_"):
        # Free allocated memory, if any
        self.estimators_ = []

    n_more_estimators = self.n_estimators - len(self.estimators_)

    if n_more_estimators < 0:
        raise ValueError(
            "n_estimators=%d must be larger or equal to "
            "len(estimators_)=%d when warm_start==True"
            % (self.n_estimators, len(self.estimators_))
        )

    elif n_more_estimators == 0:
        warn(
            "Warm-start fitting without increasing n_estimators does not "
            "fit new trees."
        )
    else:
        if self.warm_start and len(self.estimators_) > 0:
            # We draw from the random state to get the random state we
            # would have got if we hadn't used a warm_start.
            random_state.randint(MAX_INT, size=len(self.estimators_))

        trees = [
            self._make_estimator(append=False, random_state=random_state)
            for i in range(n_more_estimators)
        ]

        trees = Parallel(
            n_jobs=self.n_jobs,
            verbose=self.verbose,
            **_joblib_parallel_args(prefer="threads"),
        )(
            delayed(_parallel_build_trees_patched)(
                t,
                self,
                X,
                y,
                sample_weight,
                i,
                len(trees),
                verbose=self.verbose,
                n_samples_bootstrap=n_samples_bootstrap,
            )
            for i, t in enumerate(trees)
        )

        # Collect newly grown trees
        self.estimators_.extend(trees)

    if self.oob_score:
        self._set_oob_score(X, y)

    if hasattr(self, "classes_") and self.n_outputs_ == 1:
        self.n_classes_ = self.n_classes_[0]
        self.classes_ = self.classes_[0]

    return self


def _parallel_build_trees_patched(
    tree,
    forest,
    X,
    y,
    class_weights_df,
    tree_idx,
    n_trees,
    verbose=0,
    n_samples_bootstrap=None,
):
    """
    Private function used to fit a single tree in parallel."""
    if verbose > 1:
        print("building tree %d of %d" % (tree_idx + 1, n_trees))

    if forest.bootstrap:
        n_samples = X.shape[0]

        indices = _generate_sample_indices(
            tree.random_state, n_samples, n_samples_bootstrap
        )

        X = X.take(indices).values
        y = np.asarray(y.take(indices).values)

        if class_weights_df is not None:
            class_weights_df = class_weights_df.take(indices).values.prod(
                axis=1
            )

    tree.fit(X, y, sample_weight=class_weights_df, check_input=True)

    return tree


def _validate_y_class_weight_patched(self, y):
    class_weights_df = None

    self.classes_ = []
    self.n_classes_ = []
    self.classes_weights_ = []

    columns = list(y.columns)

    for col_name in columns:
        col = y[col_name]

        vc = col.value_counts().sort_index()
        classes_k = vc.index.values
        classes_counts = vc.values

        class_weight = classes_counts.sum() / (len(classes_k) * classes_counts)
        y[f"encoded_{col_name}"] = col.map({v: i for i, v in enumerate(classes_k)})
        y[f"weights_{col_name}"] = col.map(
            {c: w for c, w in zip(classes_k, class_weight)}
        )

        self.classes_.append(classes_k)
        self.n_classes_.append(classes_k.shape[0])
        self.classes_weights_.append(class_weight)

    if self.class_weight == "balanced":
        class_weights_df = y[[f"weights_{col_name}" for col_name in columns]]

    y = y[[f"encoded_{col_name}" for col_name in columns]]

    if self.class_weight is not None:
        valid_presets = ("balanced", "balanced_subsample")
        if isinstance(self.class_weight, str):
            if self.class_weight not in valid_presets:
                raise ValueError(
                    "Valid presets for class_weight include "
                    '"balanced" and "balanced_subsample".'
                    'Given "%s".' % self.class_weight
                )
            if self.warm_start:
                warn(
                    'class_weight presets "balanced" or '
                    '"balanced_subsample" are '
                    "not recommended for warm_start if the fitted data "
                    "differs from the full dataset. In order to use "
                    '"balanced" weights, use compute_class_weight '
                    '("balanced", classes, y). In place of y you can use '
                    "a large enough sample of the full training set "
                    "target to properly estimate the class frequency "
                    "distributions. Pass the resulting weights as the "
                    "class_weight parameter."
                )

    return y, class_weights_df


def _accumulate_prediction_patch(predict, X, out, lock):
    prediction = predict(X, check_input=False)
    estimator = predict.__self__

    if estimator.n_outputs_ == 1:
        prediction = [prediction]
        classes = [estimator.classes_]
        num_classes = [estimator.n_classes_]
    else:
        classes = estimator.classes_
        num_classes = estimator.n_classes_
    assert len(out) == len(prediction)
    with lock:
        for i, (o, pred) in enumerate(zip(out, prediction)):
            for (class_index, pred_index) in zip(
                (int(c) for c in classes[i]), range(num_classes[i])
            ):
                o[:, class_index] += pred[:, pred_index]
```

Usage Example:
```
from unittest.mock import patch

import numpy as np
import pandas as pd
import vaex
from pytest import fixture, mark
from sklearn import datasets
from sklearn.ensemble._forest import BaseForest, RandomForestClassifier
from sklearn.metrics import accuracy_score

from rf import fit_patched, _accumulate_prediction_patch


@fixture
def iris():
    return datasets.load_iris()


def train_rf(X_train, y_train, class_weight=None):
    rf = RandomForestClassifier(
        n_estimators=10,
        oob_score=False,
        random_state=42,
        max_samples=3,
        verbose=0,
        class_weight=class_weight,
    )

    rf.fit(X_train, y_train)

    return rf


def evaluate_rf(rf, X_test, y_test):
    predicted = rf.predict(X_test)
    accuracy = accuracy_score(y_test, predicted)
    print(f"Mean accuracy score: {accuracy:.3}")

    return accuracy


@fixture
def dfs(iris):
    df = pd.DataFrame(iris.data, columns=iris.feature_names)
    # sklearn provides the iris species as integer values since this is required for classification
    # here we're just adding a column with the species names to the dataframe for visualisation
    df["species"] = np.array([iris.target_names[i] for i in iris.target])
    df["target"] = iris.target

    # shuffle the DataFrame rows
    df = df.sample(frac=1, random_state=42)
    df = vaex.from_pandas(df=df)
    train_df, test_df = df.ml.train_test_split(test_size=0.4)

    return train_df, test_df


@mark.parametrize(
    ["class_weight", "y_columns"],
    [
        (None, ["target"]),
        ("balanced", ["target"]),
    ],
)
def test_results_are_consistent(iris, dfs, class_weight, y_columns):
    train_df, test_df = dfs

    # train in-memory
    train_df_pandas = train_df.to_pandas_df()
    test_df_pandas = test_df.to_pandas_df()

    X_train_pandas = train_df_pandas[iris.feature_names]
    y_train_pandas = train_df_pandas[y_columns]

    X_test_pandas = test_df_pandas[iris.feature_names]
    y_test_pandas = test_df_pandas[y_columns]

    in_mem_rf = train_rf(X_train_pandas, y_train_pandas, class_weight=class_weight)
    in_mem_accuracy = evaluate_rf(in_mem_rf, X_test_pandas, y_test_pandas)

    # train using vaex
    train_df = vaex.from_pandas(train_df_pandas)
    X_train = train_df[iris.feature_names]
    y_train = train_df[y_columns]

    X_test = test_df[iris.feature_names].values
    y_test = test_df[y_columns]

    for column_name in X_train.columns:
        X_train[column_name] = X_train[column_name].astype("float32")

    with patch.object(BaseForest, "fit", fit_patched):
        vaex_rf = train_rf(X_train, y_train, class_weight=class_weight)

    with patch(
        "sklearn.ensemble._forest._accumulate_prediction", _accumulate_prediction_patch
    ):
        accuracy = evaluate_rf(vaex_rf, X_test, y_test)

    assert in_mem_accuracy == accuracy
```

Hopefully this is useful! I'm interested to hear your thoughs.

**Comment by JovanVeljanoski** [comment]
This is great thanks! 
I will need to have a more careful look at the code and changes, have not really look at the sklearn implementation of many of their Estimators.. But this is great and the user example is helpful! 

let me take a look in a day or so, and get back to you! 
Thanks!

**Comment by maartenbreddels** [comment]
Hi Chris,

indeed, this is great! I wonder what would be required on sklearn to be able to take in a vaex expression? We've experimented a bit with this idea in https://github.com/vaexio/vaex/pull/415 .
Also, i've taken a quick look at your code, but can you highlight where the differences are, where would sklearn make a memory copy and this code would not?

Regards,

Maarten

**Comment by Chris-hughes10** [comment]
Hi Maarten,

I think the easiest way to see the difference is to compare the `patched` methods alongside the scikit counterparts. Essentially Scikit assumes that your data is in memory from the off, so makes lots of decisions based on this and frequently uses numpy functions to create new arrays; I had to strip all of that out. In terms of implementation, the main differences are:

- I had to strip away all of the validation at the random forest level (in the `fit` method), as this requires in-memory numpy arrays. This isn't a problem, as the validation can be enabled at the decision tree level.
- When balancing the class weights in `_validate_y_class_weight`, the way that scikit calculates this requires all of the `y` labels in memory, and returns a balanced array. I replaced this with an equivalent calculation, which only requires the value counts, and return a lookup, which is applied later.
- When passing data to the underlying decision trees in the `_parallel_build_trees` method, scikit passes the whole dataset to each tree, setting the sample weight to zero for samples the tree should exclude. I modified this so that only the samples that the tree requires are loaded into memory using the `take` method. If class balancing is enabled, these samples are scaled before being passed to the tree.
- As, in scikit's implementation, each tree receives the whole dataset, every class is the same for each tree when making predictions. In the modified implementation, this is no longer the case, as it is likely that some trees will not contain examples of every class. To compensate for this, I had to create a new `_accumulate_prediction` method, which uses the information from the parent estimator to correctly set the class indices given by each tree. I had to, somewhat hackily, call `self` here as scikit annoyingly only passes a reference the estimator's `fit` method, instead of the estimator itself, when making predictions.

I spent a fair bit of time exploring this, and this approach required the fewest changes to scikit; I think you would struggle to get the scikit implementation changed to support vaex and numpy due to scikit's frequent use of the numpy array constructor. Overall, I think that it isn't a huge amount of code changed, and I endeavoured to keep the structure and flow as similar as possible to ease comparisons.

Depending on the level of effort you are willing to invest, a `cleaner` approach may be to create a new RandomForest class, to act as a wrapper over scikit's decision tree estimators directly; using the modified implementations as a base. Whilst I'm confident that the design can be streamlined significantly using this approach, this does mean a lot more code to maintain, and either the reimplementation of some tried and tested scikit functionality or introducing dependencies to hidden, low-level scikit functions. Regardless, if this is an approach you would like to take, I would be happy to help with this.

Hopefully this helps, but I really think comparing the implemented functions alongside the originals for 10 minutes will provide the most clarity.

Keep up the awesome work on Vaex by the way, I'm a big fan!

Chris
